### PR TITLE
mgr/snap_schedule: restart old schedules

### PR DESCRIPTION
**Pull request by mchangir**
Old schedules were not picked up from database when mgr was restarted.
Restart old schedules on mgr restart.

Fixes: https://tracker.ceph.com/issues/54052
Signed-off-by: Milind Changire <mchangir@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by vshankar** [comment]
> Old schedules were not picked up from database when mgr was restarted.
> Restart old schedules on mgr restart.

possible to add a test case validating this fix?

**Comment by mchangir** [comment]
> > Old schedules were not picked up from database when mgr was restarted.
> > Restart old schedules on mgr restart.
> 
> possible to add a test case validating this fix?

Can the `mgr` be restarted from within the teuthology test ?

**Comment by vshankar** [comment]
> > > Old schedules were not picked up from database when mgr was restarted.
> > > Restart old schedules on mgr restart.
> > 
> > 
> > possible to add a test case validating this fix?
> 
> Can the `mgr` be restarted from within the teuthology test ?

check ceph_manager.py:revive_mgr()

File: src/pybind/mgr/snap_schedule/fs/schedule_client.py [hunk 1]
```diff
@@ -155,6 +155,14 @@ def __init__(self, mgr: Any) -> None:
         self.active_timers: Dict[Tuple[str, str], List[Timer]] = {}
         self.conn_lock: Lock = Lock()  # lock to protect add/lookup db connections
 
+        # restart old schedules
```
Comment (nmshelke) [inline src/pybind/mgr/snap_schedule/fs/schedule_client.py:158]: Is comment at line 149 related to this code fix?
```
        # TODO maybe iterate over all fs instance in fsmap and load snap dbs?
        #
        # Each db connection is now guarded by a Lock; this is required to
        # avoid concurrent DB transactions when more than one paths in a
        # file-system are scheduled at the same interval eg. 1h; without the
        # lock, there are races to use the same connection, causing  nested
        # transactions to be aborted
```

Comment (mchangir) [reply src/pybind/mgr/snap_schedule/fs/schedule_client.py:158]: @nmshelke yup, that's right
that `TODO` was a reminder to fix this exact issue
thanks for pointing it out
now that this PR addresses the `TODO`, I have removed the `TODO` comment
I'll update the PR in a bit

**Review by kotreshhr — APPROVED** [review]
LGTM

File: src/pybind/mgr/snap_schedule/module.py [hunk 2]
```diff
@@ -35,7 +35,15 @@ class Module(MgrModule):
...
-        self.client = SnapSchedClient(self)
+        self.client = SnapSchedClient(self, self.get_all_filesystems())
+
+    def get_all_filesystems(self) -> List[str]:
+        fs_list: List[str] = []
+        fs_map = self.get('fs_map')
+        if fs_map['filesystems']:
+            for fs in fs_map['filesystems']:
+                fs_list.append(fs['mdsmap']['fs_name'])
+        return fs_list
```
[the remark this answers is not in the document]
Comment (mchangir) [reply src/pybind/mgr/snap_schedule/module.py]: ack ... did the change

**Review by vshankar — APPROVED** [review]
Looks good.

@mchangir Please run it through teuthology w/ `--filter snap-schedule` and link the run here.

File: src/pybind/mgr/snap_schedule/module.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 import errno
 import json
 import sqlite3
-from typing import Any, Dict, Optional, Tuple
+from typing import Any, Dict, Optional, Tuple, List
```
Comment (joscollin) [inline src/pybind/mgr/snap_schedule/module.py:9]: missed to undo this import. Otherwise, looks good to me.

Comment (mchangir) [reply src/pybind/mgr/snap_schedule/module.py]: removed reference to unused import `List`

**Comment by mchangir** [comment]
> Looks good.
> 
> @mchangir Please run it through teuthology w/ `--filter snap-schedule` and link the run here.

@vshankar [Teuthology run using `--filter snap-schedule` with latest code](https://pulpito.ceph.com/mchangir-2022-03-01_08:24:31-fs-wip-mchangir-mgr-snap_schedule-restart-old-schedules-on-startup-distro-default-smithi/)

3 failed jobs in this run but with known and unrelated issue:

    cluster [WRN] Health check failed: Reduced data availability: 2 pgs peering (PG_AVAILABILITY)" in cluster log

I've also verified that all the snap-schedule tests have passed for all the jobs.

**Comment by vshankar** [comment]
> I've also verified that all the snap-schedule tests have passed for all the jobs.

Could you please share the link here?

**Comment by mchangir** [comment]
> > I've also verified that all the snap-schedule tests have passed for all the jobs.
> 
> Could you please share the link here?

https://pulpito.ceph.com/mchangir-2022-03-01_08:24:31-fs-wip-mchangir-mgr-snap_schedule-restart-old-schedules-on-startup-distro-default-smithi/

**Comment by vshankar** [comment]
> > > I've also verified that all the snap-schedule tests have passed for all the jobs.
> > 
> > 
> > Could you please share the link here?
> 
> https://pulpito.ceph.com/mchangir-2022-03-01_08:24:31-fs-wip-mchangir-mgr-snap_schedule-restart-old-schedules-on-startup-distro-default-smithi/

Thanks @mchangir

(not sure why I didn't get an email notification from github after an update :/)

**Comment by vshankar** [comment]
jenkins test api

**Comment by mchangir** [comment]
jenkins test make check

**Comment by mchangir** [comment]
jenkins render docs

**Comment by vshankar** [comment]
jenkins retest this please

**Comment by vshankar** [comment]
jenkins retest this please

**Review by vshankar — APPROVED** [review]

**Comment by vshankar** [comment]
jenkins render docs

**Comment by vshankar** [comment]
jenkins retest this please

**Comment by vshankar** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2022-Mar-21

**Comment by vshankar** [comment]
jenkins test make check

**Pull request merged**